_posts/2024-08-14-running-a-hsm-realm.md

```diff
@@ -0,0 +1,221 @@
+
+In our testing a 5 HSM cluster (with 3 replication groups) should be able to
+perform approximately 240 recover operations a second. This throughput drops off
+slightly as the size of the merkle tree increases. We observed a 5% drop in
+throughput going from a tree with 10M leaves to a tree with 500M leaves. Scaling
+to larger throughput will require more HSMs and replication groups.
+
+Improvements to overall availability can be achieved by running multiple
+independent realms with a N of M client configuration. e.g. 2 HSM realms and 1
+software realm with a 2 of 3 client configuration.
+
+## Google Cloud Platform
+
+The GCP services Bigtable, Secret Manager and Cloud Pub/Sub are used by the
+realm services. Credentials to access these will be needed for all hosts
+(regardless of role) in the cluster. Bigtable must be configured for a single
+zone as multi-zone configurations change the [semantics of write
+operations](https://cloud.google.com/bigtable/docs/replication-overview#consistency-model).
+
+## HSM Initialization
+
+For test clusters the `entrust_ops` tool that is part of the [HSM
+repo](https://github.com/juicebox-systems/juicebox-hsm-realm) can be used to
+initialize the NVRAM, create the Security World and the relevant realm keys. The
+[`entrust_hsm/README.md`](https://github.com/juicebox-systems/juicebox-hsm-realm/blob/main/entrust_hsm/README.md)
+has more information about this process. Unless you've purchased remote card
+readers physical access to the HSMs will be required during this process in
+order to connect the card reader and swap cards at the appropriate times.
+
+For production clusters a more rigorous key ceremony should be performed before
+the HSMs are installed in the servers. The key ceremony allows external parties
+to verify that the realm is running the expected HSM software and that only that
+version of the software has access to the keys. The
+[`ceremony`](https://github.com/juicebox-systems/ceremony) repo contains details
+of a possible key ceremony structure and process. The result of the ceremony
+includes a set of files called a Security World, a set of initialized HSMs and
+the signed software. The HSMs should then be installed in their production
+servers, and the security world files copied to the relevant location (typically
+/opt/nfast/kmdata/local) on each HSM host server.
+
+## Running the Agent
+
+On each HSM host (aka Agent) server the `entrust_agent` executable and the
+signed software files will need to be deployed in addition to the Entrust
+security world software. At this point the agent should be runnable, the logging
+will indicate that it successfully started the HSM or report some error
+condition. (typically GCP permissions). See the
+[`entrust_hsm/README.md`](https://github.com/juicebox-systems/juicebox-hsm-realm/blob/main/entrust_hsm/README.md)
+for an example.
+
+## Load Balancers
+
+Clients send their requests to the realm over HTTPS to the load balancer
```

Comment:
Since we talk about needing the agent executables above, maybe we should talk about the load balancer executables here too?

Reply:
I wonder if it's even just useful to say like, you build it like X and here's the arguments it can take (just including what the --help outputs), and let them go from there?

Reply:
I added a separate section on the build. The --help output is giant as there are a lot of optional arguments, and i don't think it's worth getting into the weeds too much. The TLS key/cert are the main things that need some thinking about.